src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs

```diff
@@ -168,24 +173,23 @@ private async Task CreateBlocksAsync()
             }
         }
 
-        /// <summary>
-        /// Doesn't let the node progress until the given time slot.
-        /// </summary>
-        /// <param name="myTimestamp">The next time slot this node should mine at.</param>
-        protected virtual async Task WaitUntilWeCanMineAsync(uint myTimestamp)
+        private async Task<uint> WaitUntilCanMineAsync()
         {
-            int waitingTime = this.GetWaitingTimeInSeconds(myTimestamp);
+            while (!this.cancellation.IsCancellationRequested)
+            {
+                uint timeNow = (uint)this.dateTimeProvider.GetAdjustedTimeAsUnixTimestamp();
 
-            this.logger.LogInformation("Waiting {0} seconds until block can be mined.", waitingTime);
+                uint myTimestamp = this.slotsManager.GetMiningTimestamp(timeNow);
 
-            // Accounts for different IDateTimeProvider implementations. The standard implementation would progress
-            // after one iteration, but for other cases where we are emulating time, this may loop more than once.
-            while (waitingTime > 0)
-            {
-                await this.TaskDelayAsync(waitingTime * 1000, this.cancellation.Token).ConfigureAwait(false);
+                int estimatedWaitingTime = (int)(myTimestamp - timeNow) - 1;
 
-                waitingTime = this.GetWaitingTimeInSeconds(myTimestamp);
+                if (estimatedWaitingTime > 0)
```

Comment:
you could get rid of `else` if you reverse the if check

Reply:
fixed